core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1541,12 +1541,21 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     map(Some(_): Option[O2]).hold(None)
 
   /** Like [[hold]] but does not require an initial value. The signal is not emitted until the initial value is emitted from this stream */
-  def hold1[F2[x] >: F[x]: Concurrent, O2 >: O]: Stream[F2, Signal[F2, O2]] =
-    this.pull.uncons1.flatMap {
-      case Some((o, tail)) => tail.hold[F2, O2](o).underlying
-      case None            => Pull.done
+  def hold1[F2[x] >: F[x]: Concurrent, O2 >: O]: Stream[F2, Signal[F2, O2]] = {
+    def go(signal: Deferred[F2, Signal[F2, O2]]) = this.pull.uncons1.flatMap {
+      case Some((o, tail)) =>
+        Pull.eval(SignallingRef.of[F2, O2](o).flatTap(signal.complete(_))).flatMap { ref =>
+          tail.foreach(ref.set(_)).underlying
+        }
+
+      case None => Pull.done
     }.streamNoScope
 
+    Stream.eval(Deferred[F2, Signal[F2, O2]]).flatMap { signal =>
+      Stream.eval(signal.get).concurrently(go(signal))
```

Comment:
This will result in `Stream.never` if the signal never gets completed, e.g. if `this` is empty.

Reply:
I'm not sure. An empty stream is also an option. That's what the original implementation did afaict.

Reply:
Ah I see, we are not concerned with the case `Stream.never.hold1` (which never emits but never ends) but with the case `Stream.empty.hold1`. So there should be an `onFinalize` or something.

Reply:
Thanks, the fix was even easier than I realized, no need for `onFinalize` :)